File: ruby/jelly/solver.rb
require_relative 'pqueue'
require 'thread'
require 'etc'

module Jelly
    class Solver
        attr_reader :check_count

        class SearchNode
            attr_reader :stage, :key, :step, :f_score
            def initialize(stage, key, step, use_bfs)
                @stage = stage
                @key = key
                @step = step
                @f_score = use_bfs ? 0 : (stage.distance || 0) + step
            end
        end

        class ThreadSafeQueue
            def initialize
                @q = []
                @mutex = Mutex.new
                @cond = ConditionVariable.new
            end

            def push(val)
                @mutex.synchronize do
                    @q << val
                    @cond.signal
                end
            end
            alias << push

            def pop(timeout = nil)
                @mutex.synchronize do
                    if timeout
                        limit = Time.now + timeout
                        while @q.empty?
                            remaining = limit - Time.now
                            break if remaining <= 0
                            @cond.wait(@mutex, remaining)
                        end
                    else
                        while @q.empty?
                            @cond.wait(@mutex)
                        end
                    end
                    @q.empty? ? nil : @q.shift
                end
            end

            def size
                @mutex.synchronize { @q.size }
            end

            def empty?
                @mutex.synchronize { @q.empty? }
            end
        end

        class ThreadSafePriorityQueue
            def initialize(&block)
                @pq = PriorityQueue.new(&block)
                @mutex = Mutex.new
                @cond = ConditionVariable.new
            end

            def push(val)
                @mutex.synchronize do
                    @pq << val
                    @cond.signal
                end
            end
            alias << push

            def pop(timeout = nil)
                @mutex.synchronize do
                    if timeout
                        limit = Time.now + timeout
                        while @pq.empty?
                            remaining = limit - Time.now
                            break if remaining <= 0
                            @cond.wait(@mutex, remaining)
                        end
                    else
                        while @pq.empty?
                            @cond.wait(@mutex)
                        end
                    end
                    @pq.empty? ? nil : @pq.shift
                end
            end

            def size
                @mutex.synchronize { @pq.size }
            end

            def empty?
                @mutex.synchronize { @pq.empty? }
            end
        end

        def initialize(no_prune: false, use_bfs: false, quiet: false, parallel: false)
            @no_prune = no_prune
            @use_bfs = use_bfs
            @quiet = quiet
            @parallel = parallel
        end

        def solve(stage, &block)
            if @parallel
                solve_parallel(stage, &block)
            else
                solve_sequential(stage, &block)
            end
        end

        def solve_sequential(stage, &block)
            detect_constraint(stage) unless @no_prune

            stage.distance = stage.estimate_distance() unless @use_bfs
            stage.freeze()
            key = stage.node_key()
            if @use_bfs
                que = []
            else
                que = PriorityQueue.new() {|x, y| x[0].distance + x[2] <=> y[0].distance + y[2]}
            end
            que << [stage, key, 0]
            nodes = {}
            nodes[key] = [nil, nil]
            check_count = 0

            until que.empty?
                stage, key, step = que.shift
                check_count += 1

                unless @quiet
                    $stderr.print "\rCheck=#{check_count}, left=#{que.size}\x1b[0K"
                end
                next if !@no_prune && unsolvable?(stage)
                if stage.solved?()
                    @check_count = check_count
                    moves = extract_moves(nodes, key)
                    if block.nil? || block.call(moves)
                        return moves
                    end
                end

                enumerate_next(stage) do |next_stage, move|
                    next_key = next_stage.node_key()
                    unless nodes.has_key?(next_key)
                        nodes[next_key] = [key, move]
                        next_stage.distance = next_stage.estimate_distance() unless @use_bfs
                        next_stage.freeze()
                        que << [next_stage, next_key, step + 1]
                    end
                end
            end
            @check_count = check_count
            return nil
        end

        def solve_parallel(stage, &block)
            detect_constraint(stage) unless @no_prune

            stage.distance = stage.estimate_distance() unless @use_bfs
            stage.freeze()
            key = stage.node_key()

            if @use_bfs
                que = ThreadSafeQueue.new
            else
                que = ThreadSafePriorityQueue.new {|x, y| x.f_score <=> y.f_score}
            end

            que << SearchNode.new(stage, key, 0, @use_bfs)

            nodes = {}
            nodes_mutex = Mutex.new
            nodes[key] = [nil, nil]

            check_count = 0
            check_mutex = Mutex.new

            active_workers = 0
            workers_mutex = Mutex.new

            goal_stage = nil
            goal_key = nil
            goal_mutex = Mutex.new

            num_threads = Etc.respond_to?(:nprocessors) ? Etc.nprocessors : 4
            threads = []

            num_threads.times do
                threads << Thread.new do
                    while true
                        break if goal_mutex.synchronize { goal_stage }

                        node = que.pop(0.01)

                        if node.nil?
                            is_done = false
                            workers_mutex.synchronize do
                                if active_workers == 0 && que.empty?
                                    is_done = true
                                end
                            end
                            break if is_done
                            next
                        end

                        workers_mutex.synchronize { active_workers += 1 }
                        begin
                            current_stage = node.stage
                            current_key = node.key
                            step = node.step

                            current_check = nil
                            check_mutex.synchronize do
                                check_count += 1
                                current_check = check_count
                            end

                            if !@quiet && current_check % 1000 == 0
                                $stderr.print "\rCheck=#{current_check}, left=#{que.size}\x1b[0K"
                            end

                            next if !@no_prune && unsolvable?(current_stage)

                            if current_stage.solved?
                                goal_mutex.synchronize do
                                    goal_stage = current_stage
                                    goal_key = current_key
                                end
                                break
                            end

                            enumerate_next(current_stage) do |next_stage, move|
                                next_key = next_stage.node_key()
                                added = false
                                nodes_mutex.synchronize do
                                    unless nodes.has_key?(next_key)
                                        nodes[next_key] = [current_key, move]
                                        added = true
                                    end
                                end

                                if added
                                    next_stage.distance = next_stage.estimate_distance() unless @use_bfs
                                    next_stage.freeze()
                                    que.push(SearchNode.new(next_stage, next_key, step + 1, @use_bfs))
                                end
                            end
                        ensure
                            workers_mutex.synchronize { active_workers -= 1 }
                        end
                    end
                end
            end

            threads.each(&:join)

            @check_count = check_mutex.synchronize { check_count }
            final_goal = goal_mutex.synchronize { goal_stage }
            final_key = goal_mutex.synchronize { goal_key }

            if final_goal
                moves = extract_moves(nodes, final_key)
                if block.nil? || block.call(moves)
                    return moves
                end
            end

            return nil
        end

        def extract_moves(nodes, key)
            moves = []
            while true
                prev_key, move = nodes[key]
                break unless prev_key
                moves << move
                key = prev_key
            end
            return moves.reverse!
        end

        def enumerate_next(stage, &block)
            skips = Set.new()
            stage.jellies.each do |jelly|
                next if skips.include?(jelly)
                2.times do |j|
                    dx = j * 2 - 1
                    updated = Solver.move_jelly(stage, jelly, dx)
                    unless updated.nil?
                        top = jelly.shape.positions.first
                        move = [jelly.x + top[0], jelly.y + top[1], dx]
                        block.call(updated, move)
                    end
                end

                unless jelly.link_next.nil?
                    other = jelly
                    skips.add(other) while (other = other.link_next) != jelly
                end
            end
        end

        def self.move_jelly(stage, jelly, dx)
            updated = stage.can_move?(jelly, dx, 0)
            return nil if updated.nil?

            while true
                fall_info = nil
                while (fall_info = updated.free_fall(fall_info))
                    ;
                end
                updated.merge_jellies()
                unless updated.apply_hiddens()
                    break
                end
            end
            return updated
        end

        def unsolvable?(stage)
            constraints = @constraints
            up = 0
            unless stage.hiddens.nil?
                constraints = @constraints.dup() unless @constraints.nil?
                stage.hiddens.each do |hidden|
                    up += 1 if hidden[:dy] <= 0
                    next if !hidden[:jelly].nil? || hidden[:link]
                    y, color, dy = hidden.values_at(:y, :color, :dy)
                    y += dy
                    next if y >= @maxy
                    if constraints.nil? || !constraints.has_key?(color) || constraints[color] > y
                        constraints = {} if constraints.nil?
                        constraints[color] = y
                    end
                end
            end
            return false if constraints.nil?

            color_heights = Hash.new(0)
            stage.jellies.each do |jelly|
                next if jelly.color == BLACK || jelly.locked || !constraints.has_key?(jelly.color)
                color_heights[jelly.color] += jelly.shape.h
            end

            stage.jellies.each do |jelly|
                next if jelly.color == BLACK || jelly.locked || !constraints.has_key?(jelly.color)
                return true if jelly.y - color_heights[jelly.color] - up >= constraints[jelly.color]
            end
            return false
        end

        def detect_constraint(stage)
            constraints = {}

            height = stage.height
            width = stage.width
            maxy = height - 1
            while true
                wall_line = stage.wall_lines[maxy]
                unless wall_line == (1 << width) - 1
                    vacant = (1..(width - 2)).any? do |x|
                        if (wall_line & (1 << x)) != 0
                            false
                        else
                            jelly = stage.jellies.find {|it| it.occupy_position?(x, maxy)}
                            jelly.nil? || !jelly.locked
                        end
                    end
                    break if vacant
                end
                maxy -= 1
            end

            stage.jellies.sort do |a, b|
                a.y != b.y ? a.y <=> b.y : a.x <=> b.x
            end.each do |jelly|
                break if jelly.y >= maxy
                next if jelly.color == BLACK || !jelly.locked
                y = jelly.y + jelly.shape.h

                # 下面が壁に覆われてたら下からつなげられないので、その分を補正
                walls = (0...jelly.shape.w).all? do |dx|
                    stage.wall_lines[y] & (1 << (jelly.x + dx)) != 0
                end
                y -= 1 if walls

                if !constraints.has_key?(jelly.color) || constraints[jelly.color] > y
                    constraints[jelly.color] = y
                end
            end

            stage.hiddens&.each do |hidden|
                next unless hidden[:jelly].nil? && hidden[:link]
                y, color, dy = hidden.values_at(:y, :color, :dy)
                y += dy
                next if y >= maxy
                if !constraints.has_key?(color) || constraints[color] > y
                    constraints[color] = y
                end
            end

            @constraints = constraints.size > 0 ? constraints : nil
            @maxy = maxy
        end
    end
end

File: java/jelly/Solver.java
package jelly;

import java.util.*;

public class Solver {
    private final boolean noPrune;
    private final boolean useBfs;
    private final boolean quiet;
    private int checkCount;
    private Map<Character, Integer> constraints;
    private int maxy;

    public Solver(boolean noPrune, boolean useBfs, boolean quiet) {
        this.noPrune = noPrune;
        this.useBfs = useBfs;
        this.quiet = quiet;
        this.checkCount = 0;
    }

    public int getCheckCount() {
        return checkCount;
    }

    private static class SearchNode {
        final Stage stage;
        final int step;
        final int fScore;

        SearchNode(Stage stage, int step, boolean useBfs) {
            this.stage = stage;
            this.step = step;
            this.fScore = useBfs ? 0 : stage.distance + step;
        }
    }

    private static class NodeHistory {
        final Stage prevStage;
        final Move move;
        NodeHistory(Stage prevStage, Move move) {
            this.prevStage = prevStage;
            this.move = move;
        }
    }

    public List<Move> solve(Stage stage) {
        if (!noPrune) {
            detectConstraint(stage);
        }

        if (!useBfs) {
            stage.distance = stage.estimateDistance();
        }
        stage.freeze();

        Queue<SearchNode> que;
        if (useBfs) {
            que = new LinkedList<>();
        } else {
            que = new PriorityQueue<>(Comparator.comparingInt(a -> a.fScore));
        }

        que.add(new SearchNode(stage, 0, useBfs));

        Map<Stage, NodeHistory> nodes = new HashMap<>();
        nodes.put(stage, new NodeHistory(null, null));

        int checkCount = 0;

        while (!que.isEmpty()) {
            SearchNode node = que.poll();
            Stage currentStage = node.stage;
            int step = node.step;
            checkCount++;

            if (!quiet) {
                System.err.print("\rCheck=" + checkCount + ", left=" + que.size() + "\033[0K");
                System.err.flush();
            }

            if (!noPrune && unsolvable(currentStage)) {
                continue;
            }

            if (currentStage.solved()) {
                this.checkCount = checkCount;
                if (!quiet) {
                    System.err.print("\rSolved!\033[0K\n");
                    System.err.flush();
                }
                return extractMoves(nodes, currentStage);
            }

            final int nextStep = step + 1;
            enumerateNext(currentStage, (nextStage, move) -> {
                if (!nodes.containsKey(nextStage)) {
                    nodes.put(nextStage, new NodeHistory(currentStage, move));
                    if (!useBfs) {
                        nextStage.distance = nextStage.estimateDistance();
                    }
                    nextStage.freeze();
                    que.add(new SearchNode(nextStage, nextStep, useBfs));
                }
            });
        }

        this.checkCount = checkCount;
        return null;
    }

    private List<Move> extractMoves(Map<Stage, NodeHistory> nodes, Stage goalStage) {
        List<Move> moves = new ArrayList<>();
        Stage curr = goalStage;
        while (true) {
            NodeHistory history = nodes.get(curr);
            if (history == null || history.prevStage == null) {
                break;
            }
            moves.add(history.move);
            curr = history.prevStage;
        }
        Collections.reverse(moves);
        return moves;
    }

    public interface NextStageCallback {
        void onNextStage(Stage nextStage, Move move);
    }

    public void enumerateNext(Stage stage, NextStageCallback callback) {
        Set<Jelly> skips = new HashSet<>();
        for (Jelly jelly : stage.jellies) {
            if (skips.contains(jelly)) {
                continue;
            }
            for (int j = 0; j < 2; j++) {
                int dx = j * 2 - 1;
                Stage updated = moveJelly(stage, jelly, dx);
                if (updated != null) {
                    Point top = jelly.shape.getPositions().get(0);
                    Move move = new Move(jelly.x + top.x(), jelly.y + top.y(), dx);
                    callback.onNextStage(updated, move);
                }
            }

            if (jelly.linkNext != null) {
                Jelly other = jelly;
                while (true) {
                    skips.add(other);
                    other = other.linkNext;
                    if (other == jelly) {
                        break;
                    }
                }
            }
        }
    }

    public static Stage moveJelly(Stage stage, Jelly jelly, int dx) {
        Stage updated = stage.canMove(jelly, dx, 0);
        if (updated == null) {
            return null;
        }

        while (true) {
            Stage.FallInfo fallInfo = null;
            while (true) {
                Stage.FallInfo nextFall = updated.freeFall(fallInfo);
                if (nextFall == null) {
                    break;
                }
                fallInfo = nextFall;
            }
            updated.mergeJellies();
            if (updated.applyHiddens() == null) {
                break;
            }
        }
        return updated;
    }

    private void detectConstraint(Stage stage) {
        Map<Character, Integer> constraints = new HashMap<>();

        int height = stage.getHeight();
        int width = stage.getWidth();
        int maxy = height - 1;

        while (true) {
            int wallLine = stage.wallLines[maxy];
            if (wallLine != (1 << width) - 1) {
                boolean vacant = false;
                for (int x = 1; x <= width - 2; x++) {
                    if ((wallLine & (1 << x)) != 0) {
                        continue;
                    }
                    final int tx = x;
                    final int ty = maxy;
                    Jelly jelly = stage.jellies.stream()
                        .filter(it -> it.occupyPosition(tx, ty))
                        .findFirst()
                        .orElse(null);
                    if (jelly == null || !jelly.locked) {
                        vacant = true;
                        break;
                    }
                }
                if (vacant) {
                    break;
                }
            }
            maxy--;
        }

        List<Jelly> sortedJellies = new ArrayList<>(stage.jellies);
        Collections.sort(sortedJellies);

        for (Jelly jelly : sortedJellies) {
            if (jelly.y >= maxy) {
                break;
            }
            if (jelly.color == Const.BLACK || !jelly.locked) {
                continue;
            }
            int y = jelly.y + jelly.shape.getH();

            boolean walls = true;
            for (int dx = 0; dx < jelly.shape.getW(); dx++) {
                if ((stage.wallLines[y] & (1 << (jelly.x + dx))) == 0) {
                    walls = false;
                    break;
                }
            }
            if (walls) {
                y -= 1;
            }

            if (!constraints.containsKey(jelly.color) || constraints.get(jelly.color) > y) {
                constraints.put(jelly.color, y);
            }
        }

        if (stage.hiddens != null) {
            for (Hidden hidden : stage.hiddens) {
                if (hidden.jelly == null && hidden.link) {
                    int y = hidden.y + hidden.dy;
                    if (y >= maxy) continue;
                    if (!constraints.containsKey(hidden.color) || constraints.get(hidden.color) > y) {
                        constraints.put(hidden.color, y);
                    }
                }
            }
        }

        this.constraints = constraints.isEmpty() ? null : constraints;
        this.maxy = maxy;
    }

    private boolean unsolvable(Stage stage) {
        Map<Character, Integer> consts = this.constraints;
        int up = 0;
        if (stage.hiddens != null) {
            if (this.constraints != null) {
                consts = new HashMap<>(this.constraints);
            } else {
                consts = new HashMap<>();
            }
            for (Hidden hidden : stage.hiddens) {
                if (hidden.dy <= 0) {
                    up++;
                }
                if (hidden.jelly != null || hidden.link) {
                    continue;
                }
                int y = hidden.y + hidden.dy;
                if (y >= this.maxy) {
                    continue;
                }
                if (!consts.containsKey(hidden.color) || consts.get(hidden.color) > y) {
                    consts.put(hidden.color, y);
                }
            }
        }

        if (consts == null) {
            return false;
        }

        Map<Character, Integer> colorHeights = new HashMap<>();
        for (Jelly jelly : stage.jellies) {
            if (jelly.color == Const.BLACK || jelly.locked || !consts.containsKey(jelly.color)) {
                continue;
            }
            colorHeights.put(jelly.color, colorHeights.getOrDefault(jelly.color, 0) + jelly.shape.getH());
        }

        for (Jelly jelly : stage.jellies) {
            if (jelly.color == Const.BLACK || jelly.locked || !consts.containsKey(jelly.color)) {
                continue;
            }
            int h = colorHeights.getOrDefault(jelly.color, 0);
            if (jelly.y - h - up >= consts.get(jelly.color)) {
                return true;
            }
        }
        return false;
    }
}
